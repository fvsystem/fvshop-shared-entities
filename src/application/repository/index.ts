export * from './cached.repository';
export * from './cached-repository.mock';
