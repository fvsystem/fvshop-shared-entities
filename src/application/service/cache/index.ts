export * from './cache.service.interface';
export * from './cache.service.mock';
