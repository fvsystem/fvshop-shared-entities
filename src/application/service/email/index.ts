export * from './email.service.interface';
export * from './email.mock';
