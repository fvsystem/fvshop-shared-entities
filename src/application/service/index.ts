export * from './cache';
export * from './email';
export * from './hash';
export * from './jwt';
export * from './logger';
