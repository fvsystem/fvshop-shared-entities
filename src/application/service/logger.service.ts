export interface LoggerServiceInterface {
  log(message: unknown): void;
  info(message: unknown): void;
  warn(message: unknown): void;
  error(message: unknown): void;
}
