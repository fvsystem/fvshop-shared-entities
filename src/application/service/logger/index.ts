export * from './logger.service.interface';
export * from './logger.mock';
