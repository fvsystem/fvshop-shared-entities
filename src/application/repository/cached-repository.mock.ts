/* istanbul ignore file */

import { EntityMock, RepositoryInterface } from '@root/domain';
import { CacheServiceInterface } from '../service';
import { CachedRepository } from './cached.repository';

export class CachedRepositoryMock extends CachedRepository<EntityMock> {
  constructor(
    cacheService: CacheServiceInterface,
    repository: RepositoryInterface<EntityMock>
  ) {
    super(cacheService, repository);
  }
}
