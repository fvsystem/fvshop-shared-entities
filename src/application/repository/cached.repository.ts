import { Entity, RepositoryInterface, UniqueEntityId } from '@root/domain';
import { CacheServiceInterface } from '../service';

export abstract class CachedRepository<E extends Entity>
  implements RepositoryInterface<E>
{
  protected readonly repository: RepositoryInterface<E>;

  protected readonly cache: CacheServiceInterface;

  protected readonly cacheKey: string;

  constructor(
    cache: CacheServiceInterface,
    repository: RepositoryInterface<E>
  ) {
    this.cache = cache;
    this.repository = repository;
    this.cacheKey = this.constructor.name;
  }

  async findAll(): Promise<E[]> {
    const list = await this.cache.get<E[]>(`${this.cacheKey}:_list`);
    if (list) {
      return list;
    }
    const entities = this.repository.findAll();
    await this.cache.set(`${this.cacheKey}:_list`, entities);
    return entities;
  }

  async update(entity: E): Promise<void> {
    await this.repository.update(entity);
    await Promise.all([
      this.cache.set(
        `${this.cacheKey}:find_id:${entity.id.toString()}`,
        entity
      ),
      this.cache.delete(`${this.cacheKey}:_list`),
    ]);
  }

  async delete(id: string | UniqueEntityId): Promise<void> {
    await this.repository.delete(id);
    await Promise.all([
      this.cache.delete(`${this.cacheKey}:find_id:${id.toString()}`),
      this.cache.delete(`${this.cacheKey}:_list`),
    ]);
  }

  async insert(entity: E): Promise<void> {
    await this.repository.insert(entity);
    await Promise.all([
      this.cache.set(
        `${this.cacheKey}:find_id:${entity.id.toString()}`,
        entity
      ),
      this.cache.delete(`${this.cacheKey}:_list`),
    ]);
  }

  async bulkInsert(entities: E[]): Promise<void> {
    await this.repository.bulkInsert(entities);
    await Promise.all([
      ...entities.map(async (entity) => {
        this.cache.set(
          `${this.cacheKey}:find_id:${entity.id.toString()}`,
          entity
        );
      }),
      this.cache.delete(`${this.cacheKey}:_list`),
    ]);
  }

  async findById(id: string | UniqueEntityId): Promise<E> {
    const entity = await this.cache.get<E>(
      `${this.cacheKey}:find_id:${id.toString()}`
    );
    if (entity) {
      return entity;
    }
    const entityFromDB = await this.repository.findById(id);
    await this.cache.set(
      `${this.cacheKey}:find_id:${id.toString()}`,
      entityFromDB
    );
    return entityFromDB;
  }
}
