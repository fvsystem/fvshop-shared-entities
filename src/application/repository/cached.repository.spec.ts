import {
  Entity,
  NotFoundError,
  RepositoryInterface,
  UniqueEntityId,
} from '@root/domain';
import { v4 as uuid } from 'uuid';
import { CacheServiceInterface } from '../service';
import { CachedRepository } from './cached.repository';

interface FakeEntityProps {
  name: string;
}

class FakeEntity extends Entity<FakeEntityProps> {
  constructor(props: FakeEntityProps, id?: string) {
    const idValue = id || uuid();
    const uniqueEntityId = new UniqueEntityId(idValue);
    super(props, uniqueEntityId);
    this.props.name = props.name;
  }

  set name(name: string) {
    this.props.name = name;
  }

  get name(): string {
    return this.props.name;
  }
}

class FakeRepository implements RepositoryInterface<FakeEntity> {
  private entities: FakeEntity[] = [];

  async findAll(): Promise<FakeEntity[]> {
    return this.entities;
  }

  async delete(id: string | UniqueEntityId): Promise<void> {
    const idValue = id instanceof UniqueEntityId ? id.value : id;
    this.entities = this.entities.filter((e) => e.id !== idValue);
  }

  async findById(id: string | UniqueEntityId): Promise<FakeEntity> {
    const idValue = id instanceof UniqueEntityId ? id.value : id;
    const entity = this.entities.find((e) => e.id === idValue);
    if (!entity) {
      throw new NotFoundError();
    }
    return entity;
  }

  async insert(entity: FakeEntity): Promise<void> {
    this.entities.push(entity);
  }

  async bulkInsert(entities: FakeEntity[]): Promise<void> {
    this.entities.push(...entities);
  }

  async update(entity: FakeEntity): Promise<void> {
    const index = this.entities.findIndex((e) => e.id === entity.id);
    if (index === -1) {
      throw new NotFoundError();
    }
    this.entities[index] = entity;
  }
}

class FakeCacheService implements CacheServiceInterface {
  private cache: Record<string, any> = {};

  async get<T = unknown>(key: string): Promise<T | null> {
    const value = this.cache[key];
    if (value) {
      return value as T;
    }
    return null;
  }

  async set<T = unknown>(
    key: string,
    value: T,
    ttlInSeconds?: number
  ): Promise<void> {
    this.cache[key] = value;
  }

  async delete(key: string): Promise<void> {
    this.cache[key] = undefined;
  }

  async clear(): Promise<void> {
    this.cache = {};
  }
}

class FakeCachedRepository extends CachedRepository<FakeEntity> {
  constructor(
    cacheService: CacheServiceInterface,
    repository: RepositoryInterface<FakeEntity>
  ) {
    super(cacheService, repository);
  }
}

describe('CachedRepository', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  it('should cache findAll', async () => {
    const cacheService = new FakeCacheService();
    const repository = new FakeRepository();
    const findAllSpy = jest.spyOn(repository, 'findAll');
    const getSpy = jest.spyOn(cacheService, 'get');
    const cachedRepository = new FakeCachedRepository(cacheService, repository);
    const entity = new FakeEntity(
      { name: 'test' },
      '7c195903-6ef3-43ed-b181-0ccb14876e28'
    );
    await repository.insert(entity);
    const cachedEntities = await cachedRepository.findAll();
    expect(cachedEntities).toEqual([entity]);
    const cachedEntities2 = await cachedRepository.findAll();
    expect(cachedEntities2).toEqual([entity]);
    expect(findAllSpy).toHaveBeenCalledTimes(1);
    expect(getSpy).toHaveBeenNthCalledWith(2, 'FakeCachedRepository:_list');
    expect(cacheService).toMatchSnapshot();
  });

  it('should cache findById', async () => {
    const cacheService = new FakeCacheService();
    const repository = new FakeRepository();
    const findByIdSpy = jest.spyOn(repository, 'findById');
    const getSpy = jest.spyOn(cacheService, 'get');
    const cachedRepository = new FakeCachedRepository(cacheService, repository);
    const entity = new FakeEntity(
      { name: 'test' },
      '7c195903-6ef3-43ed-b181-0ccb14876e28'
    );
    await repository.insert(entity);
    const cachedEntity = await cachedRepository.findById(entity.id);
    expect(cachedEntity).toEqual(entity);
    const cachedEntity2 = await cachedRepository.findById(entity.id);
    expect(cachedEntity2).toEqual(entity);
    expect(findByIdSpy).toHaveBeenCalledTimes(1);
    expect(getSpy).toHaveBeenNthCalledWith(
      2,
      `FakeCachedRepository:find_id:${entity.id}`
    );
    expect(cacheService).toMatchSnapshot();
  });

  it('should cache insert and invalidate find_all', async () => {
    const cacheService = new FakeCacheService();
    const repository = new FakeRepository();
    const insertSpy = jest.spyOn(repository, 'insert');
    const setSpy = jest.spyOn(cacheService, 'set');
    const deleteSpy = jest.spyOn(cacheService, 'delete');
    const cachedRepository = new FakeCachedRepository(cacheService, repository);
    const entity = new FakeEntity(
      { name: 'test' },
      '7c195903-6ef3-43ed-b181-0ccb14876e28'
    );
    await cachedRepository.insert(entity);
    expect(insertSpy).toHaveBeenCalledTimes(1);
    expect(setSpy).toHaveBeenNthCalledWith(
      1,
      `FakeCachedRepository:find_id:${entity.id}`,
      entity
    );
    expect(deleteSpy).toHaveBeenNthCalledWith(1, 'FakeCachedRepository:_list');
    expect(cacheService).toMatchSnapshot();
  });

  it('should cache update and invalidate find_all', async () => {
    const cacheService = new FakeCacheService();
    const repository = new FakeRepository();
    const updateSpy = jest.spyOn(repository, 'update');
    const setSpy = jest.spyOn(cacheService, 'set');
    const deleteSpy = jest.spyOn(cacheService, 'delete');
    const cachedRepository = new FakeCachedRepository(cacheService, repository);
    const entity = new FakeEntity(
      { name: 'test' },
      '7c195903-6ef3-43ed-b181-0ccb14876e28'
    );
    await cachedRepository.insert(entity);
    entity.name = 'test2';
    await cachedRepository.update(entity);
    expect(updateSpy).toHaveBeenCalledTimes(1);
    expect(setSpy).toHaveBeenNthCalledWith(
      2,
      `FakeCachedRepository:find_id:${entity.id}`,
      entity
    );
    expect(deleteSpy).toHaveBeenNthCalledWith(1, 'FakeCachedRepository:_list');
    const cachedEntity = await cachedRepository.findById(entity.id);
    expect(cachedEntity).toEqual(entity);
    expect(cachedEntity.name).toBe('test2');
    expect(cacheService).toMatchSnapshot();
  });

  it('should cache delete and invalidate find_all', async () => {
    const cacheService = new FakeCacheService();
    const repository = new FakeRepository();
    const deleteRepoSpy = jest.spyOn(repository, 'delete');
    const setSpy = jest.spyOn(cacheService, 'set');
    const deleteSpy = jest.spyOn(cacheService, 'delete');
    const cachedRepository = new FakeCachedRepository(cacheService, repository);
    const entity = new FakeEntity(
      { name: 'test' },
      '7c195903-6ef3-43ed-b181-0ccb14876e28'
    );
    await cachedRepository.insert(entity);
    await cachedRepository.delete(entity.id);
    expect(deleteRepoSpy).toHaveBeenCalledTimes(1);
    expect(setSpy).toHaveBeenNthCalledWith(
      1,
      `FakeCachedRepository:find_id:${entity.id}`,
      entity
    );
    expect(deleteSpy).toHaveBeenNthCalledWith(1, 'FakeCachedRepository:_list');
    expect(deleteSpy).toHaveBeenNthCalledWith(
      2,
      `FakeCachedRepository:find_id:${entity.id}`
    );
    expect(() => cachedRepository.findById(entity.id)).rejects.toThrowError(
      NotFoundError
    );
    expect(cacheService).toMatchSnapshot();
  });

  it('should cache bulkInsert and invalidate find_all', async () => {
    const cacheService = new FakeCacheService();
    const repository = new FakeRepository();
    const bukInsertSpy = jest.spyOn(repository, 'bulkInsert');
    const setSpy = jest.spyOn(cacheService, 'set');
    const deleteSpy = jest.spyOn(cacheService, 'delete');
    const cachedRepository = new FakeCachedRepository(cacheService, repository);
    const entity = new FakeEntity(
      { name: 'test' },
      '7c195903-6ef3-43ed-b181-0ccb14876e28'
    );
    const entity2 = new FakeEntity(
      { name: 'test2' },
      '088f6493-f1d4-47ec-b922-7c7a6655f02e'
    );
    await cachedRepository.bulkInsert([entity, entity2]);
    expect(bukInsertSpy).toHaveBeenCalledTimes(1);
    expect(setSpy).toHaveBeenNthCalledWith(
      1,
      `FakeCachedRepository:find_id:${entity.id}`,
      entity
    );
    expect(setSpy).toHaveBeenNthCalledWith(
      2,
      `FakeCachedRepository:find_id:${entity2.id}`,
      entity2
    );
    expect(deleteSpy).toHaveBeenNthCalledWith(1, 'FakeCachedRepository:_list');
    expect(cacheService).toMatchSnapshot();
  });
});
