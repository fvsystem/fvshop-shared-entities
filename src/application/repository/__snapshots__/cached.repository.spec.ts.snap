// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CachedRepository should cache bulkInsert and invalidate find_all 1`] = `
FakeCacheService {
  "cache": Object {
    "FakeCachedRepository:_list": undefined,
    "FakeCachedRepository:find_id:088f6493-f1d4-47ec-b922-7c7a6655f02e": Object {
      "id": "088f6493-f1d4-47ec-b922-7c7a6655f02e",
      "name": "test2",
    },
    "FakeCachedRepository:find_id:7c195903-6ef3-43ed-b181-0ccb14876e28": Object {
      "id": "7c195903-6ef3-43ed-b181-0ccb14876e28",
      "name": "test",
    },
  },
  "delete": [MockFunction] {
    "calls": Array [
      Array [
        "FakeCachedRepository:_list",
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": Promise {},
      },
    ],
  },
  "set": [MockFunction] {
    "calls": Array [
      Array [
        "FakeCachedRepository:find_id:7c195903-6ef3-43ed-b181-0ccb14876e28",
        Object {
          "id": "7c195903-6ef3-43ed-b181-0ccb14876e28",
          "name": "test",
        },
      ],
      Array [
        "FakeCachedRepository:find_id:088f6493-f1d4-47ec-b922-7c7a6655f02e",
        Object {
          "id": "088f6493-f1d4-47ec-b922-7c7a6655f02e",
          "name": "test2",
        },
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": Promise {},
      },
      Object {
        "type": "return",
        "value": Promise {},
      },
    ],
  },
}
`;

exports[`CachedRepository should cache delete and invalidate find_all 1`] = `
FakeCacheService {
  "cache": Object {
    "FakeCachedRepository:_list": undefined,
    "FakeCachedRepository:find_id:7c195903-6ef3-43ed-b181-0ccb14876e28": undefined,
  },
  "delete": [MockFunction] {
    "calls": Array [
      Array [
        "FakeCachedRepository:_list",
      ],
      Array [
        "FakeCachedRepository:find_id:7c195903-6ef3-43ed-b181-0ccb14876e28",
      ],
      Array [
        "FakeCachedRepository:_list",
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": Promise {},
      },
      Object {
        "type": "return",
        "value": Promise {},
      },
      Object {
        "type": "return",
        "value": Promise {},
      },
    ],
  },
  "set": [MockFunction] {
    "calls": Array [
      Array [
        "FakeCachedRepository:find_id:7c195903-6ef3-43ed-b181-0ccb14876e28",
        Object {
          "id": "7c195903-6ef3-43ed-b181-0ccb14876e28",
          "name": "test",
        },
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": Promise {},
      },
    ],
  },
}
`;

exports[`CachedRepository should cache findAll 1`] = `
FakeCacheService {
  "cache": Object {
    "FakeCachedRepository:_list": Promise {},
  },
  "get": [MockFunction] {
    "calls": Array [
      Array [
        "FakeCachedRepository:_list",
      ],
      Array [
        "FakeCachedRepository:_list",
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": Promise {},
      },
      Object {
        "type": "return",
        "value": Promise {},
      },
    ],
  },
}
`;

exports[`CachedRepository should cache findById 1`] = `
FakeCacheService {
  "cache": Object {
    "FakeCachedRepository:find_id:7c195903-6ef3-43ed-b181-0ccb14876e28": Object {
      "id": "7c195903-6ef3-43ed-b181-0ccb14876e28",
      "name": "test",
    },
  },
  "get": [MockFunction] {
    "calls": Array [
      Array [
        "FakeCachedRepository:find_id:7c195903-6ef3-43ed-b181-0ccb14876e28",
      ],
      Array [
        "FakeCachedRepository:find_id:7c195903-6ef3-43ed-b181-0ccb14876e28",
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": Promise {},
      },
      Object {
        "type": "return",
        "value": Promise {},
      },
    ],
  },
}
`;

exports[`CachedRepository should cache insert and invalidate find_all 1`] = `
FakeCacheService {
  "cache": Object {
    "FakeCachedRepository:_list": undefined,
    "FakeCachedRepository:find_id:7c195903-6ef3-43ed-b181-0ccb14876e28": Object {
      "id": "7c195903-6ef3-43ed-b181-0ccb14876e28",
      "name": "test",
    },
  },
  "delete": [MockFunction] {
    "calls": Array [
      Array [
        "FakeCachedRepository:_list",
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": Promise {},
      },
    ],
  },
  "set": [MockFunction] {
    "calls": Array [
      Array [
        "FakeCachedRepository:find_id:7c195903-6ef3-43ed-b181-0ccb14876e28",
        Object {
          "id": "7c195903-6ef3-43ed-b181-0ccb14876e28",
          "name": "test",
        },
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": Promise {},
      },
    ],
  },
}
`;

exports[`CachedRepository should cache update and invalidate find_all 1`] = `
FakeCacheService {
  "cache": Object {
    "FakeCachedRepository:_list": undefined,
    "FakeCachedRepository:find_id:7c195903-6ef3-43ed-b181-0ccb14876e28": Object {
      "id": "7c195903-6ef3-43ed-b181-0ccb14876e28",
      "name": "test2",
    },
  },
  "delete": [MockFunction] {
    "calls": Array [
      Array [
        "FakeCachedRepository:_list",
      ],
      Array [
        "FakeCachedRepository:_list",
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": Promise {},
      },
      Object {
        "type": "return",
        "value": Promise {},
      },
    ],
  },
  "set": [MockFunction] {
    "calls": Array [
      Array [
        "FakeCachedRepository:find_id:7c195903-6ef3-43ed-b181-0ccb14876e28",
        Object {
          "id": "7c195903-6ef3-43ed-b181-0ccb14876e28",
          "name": "test2",
        },
      ],
      Array [
        "FakeCachedRepository:find_id:7c195903-6ef3-43ed-b181-0ccb14876e28",
        Object {
          "id": "7c195903-6ef3-43ed-b181-0ccb14876e28",
          "name": "test2",
        },
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": Promise {},
      },
      Object {
        "type": "return",
        "value": Promise {},
      },
    ],
  },
}
`;
